aptos-move/aptos-vm/src/sharded_block_executor/mod.rs

```diff
@@ -0,0 +1,143 @@
+    },
+    thread,
+};
+
+mod block_partitioner;
+mod executor_shard;
+
+/// A wrapper around sharded block executors that manages multiple shards and aggregates the results.
+pub struct ShardedBlockExecutor<S: StateView + Sync + Send + 'static> {
+    num_executor_shards: usize,
+    partitioner: Arc<dyn BlockPartitioner>,
+    command_txs: Vec<Sender<ExecutorShardCommand>>,
+    shard_threads: Vec<thread::JoinHandle<()>>,
+    result_rx: Receiver<(usize, Result<Vec<TransactionOutput>, Error<VMStatus>>)>,
+    phantom: PhantomData<S>,
+}
+
+pub enum ExecutorShardCommand {
+    ExecuteBlock(Vec<Transaction>),
+    Stop,
+}
+
+impl<S: StateView + Sync + Send + 'static> ShardedBlockExecutor<S> {
+    pub fn new(
+        num_executor_shards: usize,
+        num_threads_per_executor: Option<usize>,
+        state_view: Arc<S>,
+    ) -> Self {
+        assert!(num_executor_shards > 0, "num_executor_shards must be > 0");
+        let num_threads_per_executor = num_threads_per_executor.unwrap_or_else(|| {
+            (num_cpus::get() as f64 / num_executor_shards as f64).ceil() as usize
+        });
+        let (result_tx, result_rx) = std::sync::mpsc::channel();
+        let mut command_txs = vec![];
+        let mut shard_join_handles = vec![];
+        for i in 0..num_executor_shards {
+            let (transactions_tx, transactions_rx) = std::sync::mpsc::channel();
+            command_txs.push(transactions_tx);
+            shard_join_handles.push(spawn_executor_shard(
+                i,
+                num_threads_per_executor,
+                state_view.clone(),
+                transactions_rx,
+                result_tx.clone(),
+            ));
+        }
+        Self {
+            num_executor_shards,
+            partitioner: Arc::new(UniformPartitioner {}),
+            command_txs,
+            shard_threads: shard_join_handles,
+            result_rx,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Execute a block of transactions in parallel by splitting the block into num_remote_executors partitions and
+    /// dispatching each partition to a remote executor shard.
+    pub fn execute_block(
+        &self,
+        block: Vec<Transaction>,
+    ) -> Result<Vec<TransactionOutput>, Error<VMStatus>> {
+        let block_partitions = self.partitioner.partition(block, self.num_executor_shards);
+        for (i, transactions) in block_partitions.into_iter().enumerate() {
+            self.command_txs[i]
+                .send(ExecutorShardCommand::ExecuteBlock(transactions))
+                .unwrap();
+        }
+        // wait for all remote executors to send the result back and append them in order by shard id
+        // maintain a map of shard id to results and aggregate them later
+        let mut results = HashMap::new();
+        trace!("ShardedBlockExecutor Waiting for results");
+        for _ in 0..self.num_executor_shards {
+            let (shard_id, result) = self.result_rx.recv().unwrap();
```

Comment:
nit: why not hold a `Vec<Receiver>` so you don't need to allocate the hashmap to sort the results

Reply:
Good point - that was pretty stupid of me. Made the code much cleaner as well :)